drivers/nutdrv_qx.c

```diff
@@ -1838,20 +1842,45 @@ static int	armac_command(const char *cmd, char *buf, size_t buflen)
 			tmpbuf[0], tmpbuf[1], tmpbuf[2], tmpbuf[3], tmpbuf[4], tmpbuf[5],
 			tmpbuf[1], tmpbuf[2], tmpbuf[3], tmpbuf[4], tmpbuf[5]);
 
+		/*
+		 * On most tested devices (including R/2000I/PSW) this was equal to the number of
+		 * bytes returned in the buffer, but on some newer UPS (R/3000I/PF1) it was 1 more
+		 * (1 control + 5 bytes transferred and bytes_available equal to 6 instead of 5).
+		 *
+		 * Current assumption is that this is number of bytes available on the UPS side
+		 * with up to 5 (ret - 1) transferred.
+		 */
 		bytes_available = (unsigned char)tmpbuf[0] & 0x0f;
 		if (bytes_available == 0) {
 			/* End of transfer */
 			break;
 		}
 
-		memcpy(buf + bufpos, tmpbuf + 1, bytes_available);
-		bufpos += bytes_available;
+		if (bytes_available > READ_SIZE - 1) {
```

Comment:
So after all, `READ_SIZE - 1` or `ret -1`, pedantically? Currently they are checked to be same, but in the possible future? Logically?..

Reply:
I believe the READ_SIZE is more explicit. ret just checks the return value and could be even out of scope afterwards.

Reply:
Well, `ret` should say how much info was actually returned by the libusb call; if it is less than the expected READ_SIZE - the rest of `tmpbuf[]` is garbage. If it is more... oops :)

Reply:
Yes, but that is checked immediately after the read in the "context" of reading USB. I'm not fluent in libusb API though and I'm unsure how it might fail. It should not read MORE data though than passed buffer or there's really a problem.

I just assume after the read+check that the tmpbuf has data read. The data can be garbage because of other fun reasons though. This USB read could be even extracted to separate function probably (with ret being available only within)

I just added a check for turned off UPS... my current one is in this state. The USB returns 0x00 bytes in this case, but DOES return data. Never tried that before.